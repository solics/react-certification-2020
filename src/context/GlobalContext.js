import React, { createContext, useReducer } from 'react'
import { globalReducer, globalInitialState } from '../reducers'

const GlobalContext = createContext()

function GlobalContextProvider({ children }) {
	const [state, dispatch] = useReducer(globalReducer, globalInitialState)
	return (
		<GlobalContext.Provider value={[state, dispatch]}>{children}</GlobalContext.Provider>
	)
}

export { GlobalContext, GlobalContextProvider }
