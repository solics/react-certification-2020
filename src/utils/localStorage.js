export const saveArrayOnLocalStorage = (id, value) => {
	localStorage.setItem(id, JSON.stringify(value))
}
export const getArrayFromLocalStorage = id => {
	return JSON.parse(localStorage.getItem(id))
}
export const addFavoriteVideo = video => {
	const favoriteVideos = JSON.parse(localStorage.getItem('favoriteVideos'))
	if (!favoriteVideos) localStorage.setItem('favoriteVideos', JSON.stringify([video]))
	else {
		favoriteVideos.push(video)
		localStorage.setItem('favoriteVideos', JSON.stringify(favoriteVideos))
	}
}
