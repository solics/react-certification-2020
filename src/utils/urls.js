const API_KEY = process.env.REACT_APP_YT_API_KEY

export const YT_API = `https://www.googleapis.com/youtube/v3/search?part=snippet&key=${API_KEY}&videoEmbeddable=true&type=video&`
export const YT_DETAIL_API = `https://www.googleapis.com/youtube/v3/videos?part=snippet&key=${API_KEY}&videoEmbeddable=true&type=video&`
