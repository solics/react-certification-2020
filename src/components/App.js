import React from 'react'
import { Route, Switch } from 'react-router'
import { HashRouter } from 'react-router-dom'
import { GlobalContextProvider } from '../context/GlobalContext'
import Layout from './Layout'
import Home from '../pages/Home'
import GlobalStyle from '../globalStyles'
import VideoDetail from '../pages/VideoDetail'
import FavoritesVideos from '../pages/FavoritesVideos'

const App = () => {
	return (
		<GlobalContextProvider>
			<GlobalStyle />
			<HashRouter>
				<Switch>
					<Route exact path="/">
						<Layout>
							<Home />
						</Layout>
					</Route>
					<Route path="/video-detail/:videoId">
						<Layout>
							<VideoDetail />
						</Layout>
					</Route>
					<Route path="/favorites">
						<Layout>
							<FavoritesVideos />
						</Layout>
					</Route>
				</Switch>
			</HashRouter>
		</GlobalContextProvider>
	)
}

export default App
