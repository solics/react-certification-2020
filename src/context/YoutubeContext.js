import React, { createContext, useState } from 'react'
import { saveArrayOnLocalStorage } from '../utils/localStorage'

const API_KEY = process.env.REACT_APP_YT_API_KEY
const MAX_RESULTS = 20

const YoutubeContext = createContext()

const SEARCH_URL = `https://www.googleapis.com/youtube/v3/search?part=snippet&key=${API_KEY}&type=video&maxResults=${MAX_RESULTS}&`

function YoutubeContextProvider({ children }) {
	const [videos, setVideos] = useState([])
	const [loading, setLoading] = useState(false)
	const [loadingRelated, setLoadingRelated] = useState(false)
	const [errorMsg, setErrorMsg] = useState('')
	const [errorMsgRelated, setErrorMsgRelated] = useState('')
	const [videosRelated, setVideosRelated] = useState([])

	const search = term => {
		setLoading(true)
		fetch(`${SEARCH_URL}q=${term}`)
			.then(response => response.json())
			.then(data => {
				setLoading(false)
				setErrorMsg('')
				if (data?.items?.length) {
					setVideos(data.items)
					saveArrayOnLocalStorage('videos', data.items)
				} else {
					setVideos([])
					setErrorMsg(data.error.message)
				}
			})
			.catch(error => {
				setLoading(false)
				setErrorMsg(error.error.message)
				console.error(error)
			})
	}

	const getRelatedVideos = videoId => {
		setLoadingRelated(true)
		fetch(`${SEARCH_URL}relatedToVideoId=${videoId}`)
			.then(response => response.json())
			.then(data => {
				setLoadingRelated(false)
				setErrorMsgRelated('')
				if (data?.items?.length) {
					setVideosRelated(data.items)
				} else {
					setVideosRelated([])
					setErrorMsgRelated(data.error.message)
				}
			})
			.catch(error => {
				setLoadingRelated(false)
				setErrorMsgRelated(error.error.message)
				console.error(error)
			})
	}

	const defaultContext = {
		videos,
		videosRelated,
		search,
		setVideosContext: newVideos => {
			setVideos(newVideos)
		},
		getRelatedVideos,
		loading,
		loadingRelated,
		errorMsg,
		errorMsgRelated,
	}

	return (
		<YoutubeContext.Provider value={defaultContext}>{children}</YoutubeContext.Provider>
	)
}

export { YoutubeContext, YoutubeContextProvider }
