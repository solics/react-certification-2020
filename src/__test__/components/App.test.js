import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { createMemoryHistory } from 'history'
import { setupServer } from 'msw/node'
import { rest } from 'msw'
import { Route, Router } from 'react-router'
import { GlobalContextProvider } from '../../context/GlobalContext'
import App from '../../components/App'
import { YT_API } from '../../utils/urls'
import mockYT from '../../youtube-videos-mock.json'

const history = createMemoryHistory()
history.push('/')

export const renderWithRouter = Component =>
	render(
		<GlobalContextProvider>
			<Router history={history}>
				<Route>
					<Component />
				</Route>
			</Router>
		</GlobalContextProvider>
	)

const handlers = [
	rest.get(`${YT_API}q=shingeki`, (req, res, ctx) => {
		return res(ctx.json(mockYT))
	}),
]
const server = setupServer(...handlers)
beforeAll(() => {
	server.listen()
})
afterAll(() => {
	server.close()
})

describe('App', () => {
	it('App Snapshot', () => {
		const component = render(<App />)
		expect(component.container).toMatchSnapshot()
	})
	it('verifies if when I search a term it brings the list of videos and look for the title of one of them', async () => {
		server.use(
			rest.post(`${YT_API}q=shingeki`, (req, res, ctx) => {
				return res(ctx.json(mockYT))
			})
		)
		const { container } = renderWithRouter(App)
		const inputSearchBox = container.querySelector('input')
		userEvent.type(inputSearchBox, 'shingeki')
		expect(container.querySelector('input')).toHaveValue('shingeki')
		fireEvent.submit(inputSearchBox)
		expect(
			await screen.findByText(/Welcome to Wizeline Guadalajara/i)
		).toBeInTheDocument()
	})
})
