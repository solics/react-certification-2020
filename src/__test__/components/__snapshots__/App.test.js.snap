// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App Snapshot 1`] = `
<div>
  <header
    class="HeaderStyled-sc-1azer9k-0 fkXvoa"
  >
    <div
      class="HeaderLeft-sc-1azer9k-2 jIiuwi"
    >
      <button
        class="MenuButton-sc-1azer9k-1 iOJGbh"
      >
        <img
          alt="menu"
          src="menu.svg"
        />
      </button>
      <form>
        <input
          class="InputSearch-sc-9xs9n4-0 kxGDRJ"
          id="search-inp"
          placeholder="Search"
          type="text"
          value=""
        />
      </form>
    </div>
    <div
      class="HeaderRight-sc-1azer9k-3 hMBPQS"
    >
      <span
        class="IndicatorStyled-j2fz6n-1 fNjceH"
      >
        Dark Mode
      </span>
      <label
        class="SwitchToggle-j2fz6n-0 dMraMU"
      >
        <input
          type="checkbox"
        />
        <span />
      </label>
      <div
        class="Avatar-sc-1azer9k-4 esSGob"
      >
        <img
          alt="avatar"
          src="user.svg"
        />
      </div>
    </div>
  </header>
  <main
    class="MainStyled-sc-33tshk-0 fypNKp"
  >
    <div
      class="HomeStyled-sc-197bdb0-0 hhdQjV"
    >
      <h1
        class="Title-sc-197bdb0-1 kVCGVh"
      >
        Results
      </h1>
      <div
        class="VideosListStyled-sc-11o1z7z-0 bZFFBb"
      >
        <p
          class="MessageStyled-sc-11o1z7z-1 icPHbH"
        >
          There's no results
        </p>
      </div>
      
    </div>
  </main>
</div>
`;
